Remove observable boilerplate + refactors

Hazelfire opened the pull request:
This PR removes all observable related code. This is because I think I've found a better way to do the observable work without needing to add it to the library in such a clunky way. This simplifies a lot of things, including something I wanted to do which was remove duplicate code in SquiggleChart and SquiggleEditor.

OAGr commented:
@berekuk could you review this? For one, note that this might require a bit of extra work on the VS Code extension.

Hazelfire commented:
Oh yeah! It probably will need changes in the VS code plugin + cli. I'll take a look through them

Hazelfire commented:
Ok, now I'm confused,  the extension never uses initialSquiggleString (which is the only thing that changes) which means it shouldn't break

berekuk commented:
Mixing controlled and uncontrolled component styles and allowing the state to be controlled both from props and from the internal editor is a bad idea. I'm pretty sure that we shouldn't do this.

The reason is that a parent of `SquiggleEditor` which sets the prop would get out of sync with `SquiggleEditor`'s internal state. Explained in more details here: https://betterprogramming.pub/dont-mix-controlled-uncontrolled-values-in-react-e7ab191d8dc4

It'd be ok to switch to the controlled mode if `squiggleString` and `onCodeChange` are set and use uncontrolled mode initialized from `initialSquiggleString` otherwise.

(btw, let's rename `squiggleString` to `code`? `squiggleString` is too long)

Hazelfire commented:
Lol, I can rename it to `code`.

How about we have two different props, `code` and `defaultCode`. When you set the first it becomes controlled and the second it becomes uncontrolled? I think that's the pattern set by [`input` elements in React](https://bobbyhadz.com/blog/react-set-default-value-of-input)

berekuk commented:
> Ok, now I'm confused, the extension never uses initialSquiggleString (which is the only thing that changes) which means it shouldn't break

Right, I added the fully controlled mode for SquigglePlayground.

berekuk commented:
> How about we have two different props, `code` and `defaultCode`. When you set the first it becomes controlled and the second it becomes uncontrolled?

Right, that's what I mean and that's what I did with SquigglePlayground.

Also, I'm confused about why this PR allowed to remove the Observable boilerplate. We still need to set `divElement.value` for `viewof` to work, controlled or uncontrolled?

Hazelfire commented:
> Also, I'm confused about why this PR allowed to remove the Observable boilerplate. We still need to set `divElement.value` for `viewof` to work, controlled or uncontrolled?

I was planning on [using this](https://observablehq.com/@j-f1/react), and rendering the components in my [setup squiggle](https://observablehq.com/@hazelfire/squiggle) notebook. They seem to have a setup where I can just create observable components based on normal react components. It has the benefit of removing react-dom as a dependency of squiggle, and reducing all the confusing boilerplate that exists already.

berekuk commented:
Ah, so you wanted to use an explicit `onChange`. But then there's an issue with accessing the outermost `div`, since you won't have an `event` object for `SquiggleEditor`.

So you would still need to use a `ref` and probably wrap `SquiggleEditor` in `forwardRef` (or use an extra div in the observable cell?) and it seems like that would just move the complexity from squiggle-components to the observable code...

To be clear, it should be possible to use the current version of code with @j-f1/react already, and I've removed the react-dom dep in #709. `renderSquiggleWhateverToDom`functions don't do anything useful anymore.

OAGr commented:
I'm going to hold off on merging until you two can come to some conclusion.

Hazelfire commented:
> So you would still need to use a ref and probably wrap SquiggleEditor in forwardRef (or use an extra div in the observable cell?) and it seems like that would just move the complexity from squiggle-components to the observable code...


I'd imagine that @j-f1/react already creates that div? Or something like it? Yes it is pushing some complexity from squiggle-components to the observable notebook, but I feel like that's probably where it belongs. The complexity isn't an issue at all when people can just import the observable functions from @hazelfire/squiggle.

I might have a go at creating an example of this PR working in an observable notebook just to make sure that I'm right about not needing this ref business in the library. I'll also fix this controlled/uncontrolled business. Putting this back into draft.

Hazelfire commented:
Hey @berekuk , I managed to get it working with a bundle without the observable boilerplate, you can see it [here](https://observablehq.com/@hazelfire/squiggle).

I think I've addressed all the issues except that the CodeEditor doesn't seem like it's actually fully controlled when I try to control it. I'll fix that and it should be ready to go

berekuk commented:
> @berekuk : Can you review this?
Will do either today or tomorrow.

berekuk commented on packages/components/README.md:
```diff
@@ -20,7 +20,7 @@ Add to `App.js`:
 import { SquiggleEditor } from "@quri/squiggle-components";
 <SquiggleEditor
-  initialSquiggleString="x = beta($alpha, 10); x + $shift"
+  squiggleString="x = beta($alpha, 10); x + $shift"
```
should be `code`

berekuk commented on packages/components/src/components/SquiggleChart.tsx:
```diff
@@ -8,20 +8,23 @@ import {
+  /** If the result is a function, where the function domain ends */
+  diagramStop?: number;
+  /** If the result is a function, the amount of stops sampled */
+  diagramCount?: number;
```
My current impression is that we should nest props more to keep them manageable in the future, while this goes in the opposite direction.

berekuk replied:
Oh, it was for consistency with SqugigleEditor! No problem then, good idea.

berekuk commented on packages/components/src/components/SquiggleEditor.tsx:
```diff
@@ -29,113 +22,41 @@ const WrappedCodeEditor: React.FC<{
+      <WrappedCodeEditor
+        code={code}
+        setCode={(code) => {
+          if (props.onCodeChange) props.onCodeChange(code);
```
This can be shortened to `props.onCodeChange?.(code)`, btw.

Hazelfire replied:
Thank you!!! I was looking for that shorthand

berekuk commented on packages/components/src/components/SquiggleEditor.tsx:
```diff
@@ -29,113 +22,41 @@ const WrappedCodeEditor: React.FC<{
-          <SquiggleErrorAlert error={result.value} />
-        )}
-      </SquiggleContainer>
-    </div>
```
Yay for removing copy-paste!

berekuk commented on packages/components/src/components/SquiggleEditor.tsx:
```diff
@@ -145,40 +66,36 @@ export interface SquigglePartialProps {
-    [initialSquiggleString]
-  );
+  const [uncontrolledCode, setCode] = useState(defaultCode);
+  let codeProp = code ?? uncontrolledCode;
```
This is counterintuitive, I expected `codeProp` to be "code that came from props".

I don't care that much, though, but I do care about consistency - the way I named the vars in `SquigglePlayground` is slightly different and we should do it in the same way in both cases.

We should just write a custom hook, actually, it would make everything much clearer.

```typescript
const [code, setCode] = useMaybeControlledValue({
  value: code,
  defaultValue: defaultCode,
  onChange: onCodeChange,
});
```

berekuk reviewed (commented):
> Hey @berekuk , I managed to get it working with a bundle without the observable boilerplate, you can see it [here](https://observablehq.com/@hazelfire/squiggle).

I checked the code, glad that it worked out! Agree that it's better than it was done before, assuming that using your observable notebook is the "official" way to use squiggle on observable (i.e., people won't try to reimplement `cell` themselves).

berekuk commented:
This is great, definitely ready for merge.
